File: implementations/fastapi/examples/src/account/infrastructure/persistence/spending_analysis_repository.py
from datetime import datetime

from sqlalchemy import UniqueConstraint, func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import Mapped, mapped_column

from ...domain.spending_analysis import SpendingAnalysis
from ...domain.spending_analysis_repository import SpendingAnalysisRepository
from .account_repository import Base


class SpendingAnalysisModel(Base):
    """The read-model table account.analyze-monthly-spending's ETL writes to, one row per
    (account_id, analysis_month) — the unique constraint is the idempotency backstop, the
    same role as sent_statement_emails' (card_id, ...) uniqueness and
    transactions.idx_transactions_reference_id_type.
    """

    __tablename__ = "spending_analysis"
    __table_args__ = (
        UniqueConstraint("account_id", "analysis_month", name="uq_spending_analysis_account_id_analysis_month"),
    )

    analysis_id: Mapped[str] = mapped_column(primary_key=True)
    account_id: Mapped[str] = mapped_column(index=True)
    analysis_month: Mapped[str]
    total_amount: Mapped[int]
    transaction_count: Mapped[int]
    average_amount: Mapped[int]
    change_from_previous_month: Mapped[int]
    trend: Mapped[str]
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)


class SqlAlchemySpendingAnalysisRepository(SpendingAnalysisRepository):
    def __init__(self, session: AsyncSession) -> None:
        self._session = session

    async def find_analysis(self, account_id: str, analysis_month: str) -> SpendingAnalysis | None:
        stmt = select(SpendingAnalysisModel).where(
            SpendingAnalysisModel.account_id == account_id,
            SpendingAnalysisModel.analysis_month == analysis_month,
        )
        row = (await self._session.execute(stmt)).scalar_one_or_none()
        if row is None:
            return None
        return self._to_domain(row)

    async def has_analysis(self, account_id: str, analysis_month: str) -> bool:
        stmt = (
            select(func.count())
            .select_from(SpendingAnalysisModel)
            .where(
                SpendingAnalysisModel.account_id == account_id,
                SpendingAnalysisModel.analysis_month == analysis_month,
            )
        )
        count = (await self._session.execute(stmt)).scalar_one()
        return count > 0

    async def save_analysis(self, analysis: SpendingAnalysis) -> None:
        self._session.add(
            SpendingAnalysisModel(
                analysis_id=analysis.analysis_id,
                account_id=analysis.account_id,
                analysis_month=analysis.analysis_month,
                total_amount=analysis.total_amount,
                transaction_count=analysis.transaction_count,
                average_amount=analysis.average_amount,
                change_from_previous_month=analysis.change_from_previous_month,
                trend=analysis.trend,
                created_at=analysis.created_at,
            )
        )
        await self._session.flush()

    def _to_domain(self, row: SpendingAnalysisModel) -> SpendingAnalysis:
        return SpendingAnalysis(
            analysis_id=row.analysis_id,
            account_id=row.account_id,
            analysis_month=row.analysis_month,
            total_amount=row.total_amount,
            transaction_count=row.transaction_count,
            average_amount=row.average_amount,
            change_from_previous_month=row.change_from_previous_month,
            trend=row.trend,  # type: ignore[arg-type]
            created_at=row.created_at,
        )

File: implementations/go/examples/internal/infrastructure/persistence/spending_analysis_repository.go
package persistence

import (
	"context"
	"database/sql"
	"errors"
	"fmt"

	"github.com/example/account-service/internal/domain/account"
)

// AccountRepository also implements account.SpendingAnalysisRepository — the
// same "one concrete struct, several domain interfaces" idiom it already
// uses for account.Repository/account.Query, rather than a second struct —
// the spending_analysis table has no interesting relationship to model
// beyond "one row per (account_id, analysis_month)" so a dedicated
// persistence struct would add nothing.
var _ account.SpendingAnalysisRepository = (*AccountRepository)(nil)

// SaveAnalysis inserts a new spending_analysis row. AnalysisID is always a
// freshly generated one (see account.NewSpendingAnalysis), so this is a
// plain INSERT rather than an upsert — a race that slips past the
// application-level HasAnalysis precheck surfaces as a unique-constraint
// violation on (account_id, analysis_month), the same "the DB constraint is
// the last line of defense, and hitting it is a genuine error, not silently
// swallowed" behavior as transactions.idx_transactions_reference_id_type.
func (r *AccountRepository) SaveAnalysis(ctx context.Context, a *account.SpendingAnalysis) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO spending_analysis
		   (id, account_id, analysis_month, total_amount, transaction_count, average_amount, change_from_previous_month, trend, created_at)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
		a.AnalysisID, a.AccountID, a.AnalysisMonth, a.TotalAmount, a.TransactionCount, a.AverageAmount, a.ChangeFromPreviousMonth, string(a.Trend), a.CreatedAt,
	)
	if err != nil {
		return fmt.Errorf("save spending analysis: %w", err)
	}
	return nil
}

// HasAnalysis implements the idempotency precheck required by
// account.SpendingAnalysisQuery.
func (r *AccountRepository) HasAnalysis(ctx context.Context, accountID, analysisMonth string) (bool, error) {
	var count int
	if err := r.db.QueryRowContext(ctx,
		`SELECT COUNT(*) FROM spending_analysis WHERE account_id = $1 AND analysis_month = $2`,
		accountID, analysisMonth,
	).Scan(&count); err != nil {
		return false, fmt.Errorf("has spending analysis: %w", err)
	}
	return count > 0, nil
}

// FindAnalysis looks up the (accountId, analysisMonth) row, returning
// account.ErrSpendingAnalysisNotFound if none exists yet.
func (r *AccountRepository) FindAnalysis(ctx context.Context, accountID, analysisMonth string) (*account.SpendingAnalysis, error) {
	var a account.SpendingAnalysis
	var trend string
	err := r.db.QueryRowContext(ctx,
		`SELECT id, account_id, analysis_month, total_amount, transaction_count, average_amount, change_from_previous_month, trend, created_at
		 FROM spending_analysis WHERE account_id = $1 AND analysis_month = $2`,
		accountID, analysisMonth,
	).Scan(&a.AnalysisID, &a.AccountID, &a.AnalysisMonth, &a.TotalAmount, &a.TransactionCount, &a.AverageAmount, &a.ChangeFromPreviousMonth, &trend, &a.CreatedAt)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, account.ErrSpendingAnalysisNotFound
		}
		return nil, fmt.Errorf("find spending analysis: %w", err)
	}
	a.Trend = account.SpendingTrend(trend)
	return &a, nil
}
